okhttp/src/main/java/okhttp3/internal/concurrent/TaskRunner.kt

```diff
@@ -38,117 +37,222 @@ import java.util.concurrent.TimeUnit
 class TaskRunner(
   val backend: Backend = RealBackend()
 ) {
-  // All state in all tasks and queues is guarded by this.
+  private var coordinatorWaiting = false
+  private var coordinatorWakeUpAt = 0L
 
-  private var coordinatorRunning = false
-  private val activeQueues = mutableListOf<TaskQueue>()
-  private val coordinator = Runnable { coordinate() }
+  /** Queues with tasks that are currently executing their [TaskQueue.activeTask]. */
+  private val busyQueues = mutableListOf<TaskQueue>()
 
-  fun newQueue() = TaskQueue(this)
+  /** Queues not in [busyQueues] that have non-empty [TaskQueue.futureTasks]. */
+  private val readyQueues = mutableListOf<TaskQueue>()
 
-  /**
-   * Returns a snapshot of queues that currently have tasks scheduled. The task runner does not
-   * necessarily track queues that have no tasks scheduled.
-   */
-  fun activeQueues(): List<TaskQueue> {
-    synchronized(this) {
-      return activeQueues.toList()
+  private val runnable: Runnable = object : Runnable {
+    override fun run() {
+      while (true) {
+        val task = synchronized(this@TaskRunner) {
+          awaitTaskToRun()
+        } ?: return
```

Comment:
I think my biggest concern is that the delay of the time to execute a solitary task becomes the limiting time to execute a later (but soon) task on another queue.

Or am I missing something?

Do you schedule a kick for next next task time somewhere?

Reply:
Yeah, awaitTaskToRun might be misnamed. It’ll always schedule a thread if there’s something to do now (run or sleep) that competes with something to do later (run or sleep).